src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs

```diff
@@ -54,37 +54,21 @@ internal partial class SymbolTreeInfo : IChecksummedObject
         private readonly OrderPreservingMultiDictionary<int, int> _inheritanceMap;
 
         /// <summary>
-        /// Maps the name of target type name of simple extension methods to its <see cref="ExtensionMethodInfo" />.
+        /// Maps the name of receiver type name to its <see cref="ExtensionMethodInfo" />.
         /// <see cref="ParameterTypeInfo"/> for the definition of simple/complex methods.
+        /// For non-array simple types, the receiver type name would be its metadata name, e.g. "Int32".
+        /// For any array types with simple type as element, the receiver type name would be just "ElementTypeName[]", e.g. "Int32[]" for int[][,]
+        /// For non-array complex types, the receiver type name is "".
+        /// For any array types with complex type as element, the receier type name is "[]"
         /// </summary>
-        private readonly MultiDictionary<string, ExtensionMethodInfo>? _simpleTypeNameToExtensionMethodMap;
+        private readonly MultiDictionary<string, ExtensionMethodInfo>? _receiverTypeNameToExtensionMethodMap;
 
-        /// <summary>
-        /// A list of <see cref="ExtensionMethodInfo" /> for complex extension methods.
-        /// <see cref="ParameterTypeInfo"/> for the definition of simple/complex methods.
-        /// </summary>
-        private readonly ImmutableArray<ExtensionMethodInfo> _extensionMethodOfComplexType;
-
-        public bool ContainsExtensionMethod => _simpleTypeNameToExtensionMethodMap?.Count > 0 || _extensionMethodOfComplexType.Length > 0;
-
-        public ImmutableArray<ExtensionMethodInfo> GetMatchingExtensionMethodInfo(ImmutableArray<string> parameterTypeNames)
-        {
-            if (_simpleTypeNameToExtensionMethodMap == null)
-            {
-                return _extensionMethodOfComplexType;
-            }
-
-            var builder = ArrayBuilder<ExtensionMethodInfo>.GetInstance();
-            builder.AddRange(_extensionMethodOfComplexType);
+        public MultiDictionary<string, ExtensionMethodInfo>.ValueSet GetExtensionMethodInfoForReceiverType(string typeName)
+            => _receiverTypeNameToExtensionMethodMap != null
+                ? _receiverTypeNameToExtensionMethodMap[typeName]
+                : new MultiDictionary<string, ExtensionMethodInfo>.ValueSet(null, null);
```

Comment:
singleton for this?

Reply:
it's a struct

Reply:
Good to know.  a readonly static field might still be nice for clarity, but i don't care then :)